[WIP] Subgraph op

reminisce opened the pull request:
## Description ##
@zheng-da 

## Checklist ##
### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

reminisce committed:
subgraph op

Fix

reminisce committed:
Add treating whole graph as a subgraph

reminisce committed:
Add forward executor in subgraph op forward function

reminisce committed:
Add unit test

reminisce committed:
Add finding subgraphs

reminisce commented on src/operator/subgraph/partition_graph.cc:
```diff
@@ -0,0 +1,208 @@
+  //std::unordered_map<SimpleNodePtr, int> outputs;
+};
+
+void CreateSimpleGraph(const Graph& g,
```
@zheng-da Call this to create an undirected graph from `g`.

reminisce commented on src/operator/subgraph/partition_graph.cc:
```diff
@@ -0,0 +1,208 @@
+}
+
+// number of subgraphs found
+void FindSubgraphs(const Graph& g,
```
@zheng-da Call this to find all subgraphs.

reminisce committed:
Add find input/output entries of a subgraph

reminisce commented on src/operator/subgraph/partition_graph.cc:
```diff
@@ -0,0 +1,262 @@
+}
+
+// find the input entries of a subgraph
+void FindInputEntries(const Graph& g,
```
@zheng-da

reminisce commented on src/operator/subgraph/partition_graph.cc:
```diff
@@ -0,0 +1,262 @@
+}
+
+// find the output entries of a subgraph
+void FindOutputEntries(const Graph& g,
```
@zheng-da

reminisce committed:
Add get input/output entries

reminisce committed:
Add debug info

reminisce committed:
Handle corner case when a subgraph node is the last node of the graph

zheng-da commented on src/operator/subgraph/partition_graph.cc:
```diff
@@ -0,0 +1,208 @@
+
+/*!
+ *  Copyright (c) 2016 by Contributors
+ * \file quantization.cc
```
file name is incorrect.

zheng-da commented on src/operator/subgraph/partition_graph.cc:
```diff
@@ -0,0 +1,208 @@
+  simple_nodes->reserve(indexed_graph.num_nodes());
+  for (size_t nid = 0; nid < indexed_graph.num_nodes(); ++nid) {
+    SimpleNodePtr sn = SimpleNode::Create();
+    sn->node = const_cast<nnvm::Node*>(indexed_graph[nid].source);
```
this is kind of hacky. i think it's better to call DFSVisit to get nodes.

reminisce replied:
Yes, this is a quick implementation for getting the flow to run through. Will get rid of this later.

zheng-da commented on src/operator/subgraph/partition_graph.cc:
```diff
@@ -0,0 +1,208 @@
+                   const int label,
+                   const size_t snid,  // simple node id
+                   const std::vector<SimpleNodePtr>& simple_nodes,
+                   std::vector<SimpleNode*>* subgraph_nodes) {
```
can you add comments on this function?
my understanding is that this function labels nodes that connect to a starting node via both input links and output links.

zheng-da replied:
You could generalize this function by
* generalize the node selection criteria (e.g., passing a functor),
* allow users to specify following input links, output links or both.

reminisce replied:
Agree. For now I only focus on getting the flow to run through in my use case. After sorting out every detail of the implementation, we can revisit and come up with a generic interface design.

zheng-da commented on src/operator/subgraph/partition_graph.cc:
```diff
@@ -0,0 +1,208 @@
+  const size_t num_outputs = list_output_names_func(subgraph_node_ptr->attrs).size();
+  for (uint32_t i = 0; i < num_outputs; ++i) {
+    ret.outputs.emplace_back(nnvm::NodeEntry{subgraph_node_ptr, i, 0});
+  }
```
i suppose this function isn't complete yet?

reminisce committed:
Add var version for ndarray

reminisce committed:
Fix unit test

zheng-da committed:
generalize graph partitioning. (#11)

* add functions for cutting edges.

* construct subgraphs.

* generalize graph partition.

* restructure the code.

* create SubgraphOpState.

* register subgraph property.

* rename.

* address comments.

* update select API.

* rename.

* set subgraph property.

* fix bugs.

* fix bugs.

zheng-da committed:
change the interface (#13)

* remove SubgraphOperator.

* change the interface of SubgraphSelector.

reminisce committed:
Reorder input/output entries to follow original topological sort order

reminisce committed:
Make subgraph op node name unique

reminisce commented:
Graph partitioning example of replacing operators: Convolution, BatchNorm, Pooling, and Relu with subgraph operators in Resnet-152.
- Original graph
![image](https://user-images.githubusercontent.com/4978794/40871278-87e37dfc-65ed-11e8-9c40-c9a40e13aea6.png)

- Partitioned graph
![image](https://user-images.githubusercontent.com/4978794/40871268-710fa6d2-65ed-11e8-8c35-d0db8adfa80f.png)

reminisce committed:
Fix sorting entries

reminisce committed:
Allow for adjacent subgraphs

reminisce committed:
Enable partitioning algo to eliminate cycles

zheng-da committed:
fix shape/dtype/storage inference. (#15)

* update tests.

* fix shape/dtype/storage inference.

* fix.

reminisce committed:
Remove graph executor from subgraph op forward

reminisce committed:
Naive engine end to end run through

zheng-da committed:
execute subgraph operators synchronously (#16)

* execute subgraph operators correctly.

* add HasSubgraph attr.

* support subgraph op in imperative.

* rewrite running subgraph in main thread.

* execute subgraph op in imperative.

* remove need_grad.

reminisce committed:
Fix graph partitioning bug and add var version for naive engine

larroy committed:
Fixes for CI #11214 (#11217)

XiaotaoChen committed:
Support for dot(dns, csr) = dns and dot(dns, csr.T) = dns on CPU (#11113)

* implement dot(dns, csr/csr.T)=dns on cpu

* complete documentaion related to dot(dns, csr/csr.T)=dns on cpu

* support fp16 by replacing MSHADOW_SGL_DBL_TYPE_SWITCH with MSHADOW_REAL_TYPE_SWITCH

ZiyueHuang committed:
[MXNET-394] concat of CSR NDArrays on first dimension (#11024)

* concat of csr ndarrays on first dimension

* update

* update

* trigger

* CR comments

* update docs and sparse op alias

* CR comments

* update

* update

haojin2 committed:
[MXNET-420] broadcast_mul/div between csr and 1D dense on GPU (#10939)

* broadcast_mul/div between csr and 1D dense on GPU

* handle elemwise case for broadcast_mul

* add docs and comments

ZiyueHuang committed:
fix propagation of cpu shared context, issue #11160 (#11182)

* concat of csr ndarrays on first dimension

* update

* update

* trigger

* CR comments

* update docs and sparse op alias

* fix propagation of cpu shared context

* revert unneccessary files

* add unittest

reminisce committed:
Refactor

reminisce committed:
Toggle engine var version print out

reminisce committed:
Add virtual destructor to engine var

asitstands committed:
Replace the old adhoc method to iterate over gpu devices with new mx.context.num_gpus (#11227)

eric-haibin-lin committed:
[WIP] Gluon sparse block and sparse embedding (#11197)

* add sparse block

* add sparse embedding

* add doc

* lint

* remove sparseblock

reminisce committed:
Fix gpu build and add example

reminisce committed:
Fix FMutateInputs for subgraph op

kpmurali committed:
[MXNET-530] Remove install page artifacts (#11191)

* Removing the update_install method for manual artifact copying

* Deleting the static install files from artifacts directory

reminisce committed:
Add unit test for FMutateInputs attr of subgraph

ptrendx committed:
Support for data iterators returning lists of batches (#11112)

* Making list of DataBatches work

* Fix a bug in score

* Fix test_io option in image-classification example

* Fix pylint

CodingCat committed:
[MXNET-62] add test against spark integration (#10462)

* fix bug

* temp

* temp

* temp

* update

* update

* update

* remove debugging stubs

* remove unused

* stylistic fix

* fix typo

* Pulled down update to submodule_dir

* add test

* retrigger it

* sync 3rd party

eric-haibin-lin committed:
Fix a bug in sparse embedding operator (#11231)

* add sparse block

* add sparse embedding

* add doc

* lint

* remove sparseblock

* fix embedding

larroy committed:
Fix #11214 ccache compilation (#11244)

zhreshold committed:
fix loading params if ignore_extra is set (#11235)

anirudhacharya committed:
[MXNET-344] [ONNX-MXNet] Add new Operator Translations for ONNX import module (#11140)

* gluon import

* gluon tests

* shape issues.

* remove the dim_change list

* onnx backend tests

* changes to match onnx op set version 7

* fix

* lint fix

* add new folder

* fix

* fix

* rename test file

* tile topk clip

* more ops

* fix

* fix

lanking520 committed:
[MXNET-386] ongoing maintenance on NDArray (#11126)

* Important ndarray feature

* merge generic function Invoke

* Pass the Scala Style test

* add Experimental tags

* Change with NDArgs addition

* change dir for Experimental tag

* reTrigger CI

* add Symbol Macros change

* Add some workaround on NDArray

* Simplify the base part

* add changes on ND and Symbols...

* avoid vars

* add Symbol Macros

* Trigger the CI

* Trigger CI

larroy committed:
Fix build.py when CCACHE_DIR is set. (#11226)

Don't use tempdir on osx since this is not mountable by default in Docker in osx

ThomasDelteil committed:
[MXNET-534] Add examples to example list (#11225)

* Adding examples to list

* Update links add demo

* remove onnx-mxnet as it is now part of contrib

* Trigger build

* Adding Awesome-MXNet

* Trigger build

szha committed:
gpu mem pool strategy (#11041)

* use nearest power of 2 for gpu memory pool sizes

* add linear

* add test

lanking520 committed:
[MXNET-543] disable scalatest on Spark (#11264)

* disable scalatest on Spark

* Kill the entire build

* Disable test in test folder and add dummy label

wentingj committed:
[MXNET-290] MKLDNN support for model quantization (#10433)

* mkldnn support for quantization

* fix output number in graph

* update licsence

* modify Jenkinsfile

* modify Jenkinsfile

* mkldnn has no int8 fc api, excluded_sym_names includes fc for cpu

* add mkldnn uint8 pass for quantization graph

* update ut

* retrig ic

* remove no mkldnn quantization test temp

* seperate mkldnn quantization ut from gpu quantization ut

* rm dev_id check for cpu

* add mkl tests dictionary

* resolve review comments

* simplify DequantizeStorageType() logic

* simplify quantize/quantized_conv storage type logic

* Add mkldnn_OIhw4i16o4i type case (needed by int8)

* INT8 conv/pooling: share with FP32 convolution/pooling class/function

* minor indent changes

* Remove unnecessary mkldnn_quantized_pooling-inl.h

* Fix minor issue

* Fix lint

* delete duplicated data type

* fix bugs and convert requantize data to NDArray

* fix lint

* fix requantize storgetype

* fix requantize storge type

* Fix coding style comments

* Fix compile issue

* Change to use quantized_dtype option to support uint8/int8 scenarios

* fix gpu test quantization failure

* Fix indent

* fix quantized pooling param parser

* Fix imagenet_gen_qsym.py option style

* retrigger jenkins

* retrigger again

* trigger jenkins

* Resolve further comments

* share test code

* remove unnecessary test code

* add test_quantize_model for cpu

* add comments in quantize_graph_pass.cc

* jenkins

* jenkins

* improve coding style

* improve coding style

* Add naive CPU quantization test back and share quantization code between naive-CPU/MKLDNN/GPU

* rename test_quantization_cpu.py to test_quantization_mkldnn.py

* code style

* trigger

* Adjust variable naming for test quantization

* add qdtype for quantized op test case to test/bypass all cases explicitly

* change expressions to be consistent

* revert unnecessary change

piiswrong committed:
add import_ for SymbolBlock (#11127)

* add import_ for SymbolBlock

* fix

* Update block.py

* add save_parameters

* fix

* fix lint

* fix

* fix

* fix

* fix

* fix

* Update save_load_params.md

andrewfayres committed:
Scala inference memory leak fix (#11204)

* Fixes Scala memory leak (#10436)

kpmurali committed:
[MXNET-521] Add Facebook open-graph tag integration (#11155)

* Adding the first set of Facebook OG tags to layout.html

* Adding the direct link to the Facebook OG image

* Re-arranging the og description tag

* Changes to fix merge conflicts

* Adding secure_url to the og:image tag

* Adding the redundant og:image tag

azai91 committed:
[MXNET-546] Add unit test for MKLDNNSum (#11272)

* add test for mkldnnsum

* add extra comment

* fix verify

* only run if mkldnn supported

* divide size by size of type

* filter out regular arrays

* fix cond

* add verify mes

* filter views

* different outputs arrays

* move print message

* add in place test

* update copy fn

* refactor copyfrom

* use arr.copy instead of tmpmemmg

* use InitMKLDNNArray helper

* fix params

* pass correct type to copyfrom

* add print message for inplace sum

* remove copyfrom refactor

* remove redundant header

* remove extra line

* fix lint

* retrigger

piiswrong committed:
Improve hybridblock doc (#11236)

* Improve hybridblock doc

* fix

Commit:
disable testcase test_gru_bidirectional temporarily (#11273)

* disable testcase test_gru_bidirectional temporarily

* retrigger the build and see if permission denied /efs-ccache can pass

ThomasDelteil committed:
[MXNET-532] Clarify documentation of save_parameters(), load_parameters() (#11210)

* flipping official and alternative

* Update block.py

* Add new line

* fix trailing whitespaces

* clarify docs

* Update block.py

* Update block.py

* fix typo

* trailing whitespace

* Trigger build

marcoabreu committed:
Enable shared ccache and fix test (#11269)

* Enable shared ccache

* Increase cache size

* Don't fail if EFS ccache is not available

hetong007 committed:
Improve data transform for gluon data loader (#11183)

* improve transforms.Resize

* fix

* Trigger CI

* Trigger CI

* improve

* Trigger CI

* Trigger CI

* fix unittest

* keep_ratio is false by default, to keep consistency

larroy committed:
[ARM] improvements to ARMv7 based builds. (#11245)

Fix build with OpenCV 2. Native RPi build.
Openblas compilation fixes and version pinning. (@lebeg)
Disabled bundled OpenMP for cross compilation. (@lebeg)
Build logic refinements.

marcoabreu committed:
[MXNET-538] Add XUnit to python tests (#11199)

* Add XUnit to python tests

* Disable xunit for Windows

* Update indenting

* Fix python test_utils

* Rename files

* Accomodate latest changes on master

tqchen committed:
[BUILD] Remove useless additional link opt (#11300)

* [BUILD] Remove useless additional link opt

* Update tvm dep

larroy committed:
[MXNET-57] Add android64 build (#11188) (#11055)

* Fixes sources for android64
Add initial android64 build logic.
Remove pthread when linking MXNet in Android. Bionic provides built-in support for threads.
* Simplify and fix android64 build

szha committed:
leaky relu speed (#11012)

* leaky relu forward speed

* leaky relu backward speed

* fix infer shape

* fix shape

kpmurali committed:
[MXNET-541] Correcting the new install pages bugs (#11261)

* Correcting the install command for version 0.11.0

* added latest CUDA versions and link to old versions

* fixed typo

* Changing the queryString variable to be platform instead of Devices

* Fixing the unclosed HTML comment tag

marcoabreu committed:
Revert "[WIP] Do Not Merge. Static memory allocation for cached_op (#10817)" (#11311)

This reverts commit 2dbd143e4892bb9ad4aa1835c79f0046603e3531.

yzhliu committed:
[MXNET-539] Allow Scala users to specify data/label names for NDArrayIter (#11256)

* improve NDArrayIter to have Builder and ability to specifying names

piiswrong committed:
Static alloc for hybridblock (#11313)

Thanks for the contribution, this is now merged

marcoabreu committed:
Revert "Static alloc for hybridblock (#11313)" (#11318)

This reverts commit 5431e12f11fc5446f6ec2a25098b4e3b67ee7eb3.

TaoLv committed:
fix DefaultSubgraphOpMutableInputs (#17)

haojin2 committed:
bump up rtol for fp16 case of test_sgd (#11246)

ThomasDelteil committed:
[MXNET-352] Document behavior of mx.initializer.Constant (#10637)

* document behavior of Constant

* Update initializer.py

* remove trailing whitespace

* Update initializer.py

* Update initializer.py

* Trigger build

* Trigger build

KellenSunderland committed:
[MXNET-554] Add ccache reporting to CI (#11322)

yasserzamani committed:
[MXNET-540] update python gpu build from source instructions on windows (#11276)

* update python gpu build from source instructions on windows

See Also: MXNET-540

* delete duplicate entry from copy/paste in prev commit

myme5261314 committed:
Fix gperftools/jemalloc and lapack warning bug. (#11110)

* Fix gperftools/jemalloc and lapack warning bug.

* mend

congchan committed:
allow user to define unknown token symbol (#10461)

test case added

azai91 committed:
[MXNET-542] Fix mkldnn performance regression + improve test logging (#11262)

* do not create tmp memory during act

* fix order of alloc memory

* fix conditional

* fix order

* do not pass nullptr to commit

* fix comment

* do not create tmp mem unless shapes diff

* fix params

* always return in CreateMKLDNNMem

* add boilerplate for CreateMKLDNNMem test

* refactor copyfrom

* use copyfrom helper in tests

* add logs

* missing semi

* improve print msg

* target out_mem

* test copy from

* reuse verify copy

* add inplace test / use sum for test

* use assert in sum verify

* lint

* remove unused var

* fix test messsage

* out_mem can be null

* Revert "refactor copyfrom"

This reverts commit 4ab131ee41832eefb32971f82a58440736ca3417.

* add back missing var

* writeinplace explicitly returns same memory

* refactor

* only writeinplace if add and pdesc are eq

* fix comparison

* add second CreateMKLDNNMemory

* CreateMKLDNNMem accepts input

* refactor WriteTo criteria into separate method

* fix lint

* copyfrom test back

* update mldnnsum test to have diff inputs for write in place

* test in place sum with diff arrs

* revert CreateMKLDNNMem extra param change

* pass input arr param for act_forward

* remove extra header

* fix indent

* add check for writeto

* canwriteto uses ref instead of ptr

* update comments for CreateMKLDNNData

* compare input and output desc with op pdesc

* check CreateMKLDNNData does not return null

ctcyang committed:
Bring back MXNET_GPU_COPY_NTHREADS env variable (#11237)

* bring back MXNET_GPU_COPY_NTHREADS env variable

* make variable size_t for consistency

apeforest committed:
Fix the inaccruate instruction on installing nose (#11317)

* Fix the inaccruate instruction on installing nose

* Clarified the installation of nosetests and removed 'sudo'

* Update README.md

haojin2 committed:
[MXNET-380] count_include_pad argument for Avg Pooling (#11021)

* add count_include_pad argument

* add cound_include_pad in cudnn pooling and corresponding tests

* add gluon support for the new option

* switch to built-in functions for artificial NaN

* add doc for gluon

* change isAvg/getAvg to is_avg/get_avg

leezu committed:
MXNET_FORCE_ADDTAKEGRAD to disable AddTakeGradLargeBatchCaller (#11316)

* MXNET_FORCE_ADDTAKEGRAD to disable AddTakeGradLargeBatchCaller

If MXNET_FORCE_ADDTAKEGRAD is set, EmbeddingOpBackward will always use
AddTakeGrad independently of gradient input and output shape

* Read MXNET_FORCE_ADDTAKEGRAD to a static variable

eric-haibin-lin committed:
[MXNET-514] Add clip_global_norm(row_sparse_grad). Fix row_sparse_param.save(). Fix trainer init_kvstore (#11266)

* clip sparse grad. fix _reduce for rowsparse param

* fix kvstore init for local kv

* trigger

reminisce committed:
Add post processing algorithm for filtering out nodes

indhub committed:
[MXNET-547] Tutorial explaining how to use the profiler	 (#11274)

* Add first draft of profiler tutorial

* Minor changes

- Add images from web-data
- Add <!--notebook-skip-line-->

* Language corrections

* Minor changes

- Fix image URLs
- Fix formatting of output

* Minor changes

- Add download button.
- Hide profile_stats.png in notebook.

* Add tutorial to index.

* Add tutorial to tests.

* Add a note about nd.waitall()

* Remove the example build command.

Link to installation page is sufficient.

* Fix typo

* Include info about env variables related to profiling

* Add a further reading section

reminisce committed:
Refactor post processing

reminisce committed:
Change file names

KellenSunderland committed:
[MXNET-553] Restructure dockcross dockerfiles to fix caching (#11302)

* Add ccache reporting to CI

* Restructure dockcross dockerfiles to fix caching

hetong007 committed:
Add standard ResNet data augmentation for ImageRecordIter (#11027)

* add resnet augmentation

* add test

* fix scope

* fix warning

* fix lint

* fix lint

* add color jitter and pca noise

* fix center crop

* merge

* fix lint

* Trigger CI

* fix

* fix augmentation implementation

* add checks for parameters

* modify training script

* fix compile error

* Trigger CI

* Trigger CI

* modify error message

* Trigger CI

* Trigger CI

* Trigger CI

* improve script in example

* fix script

* clear code

* Trigger CI

* set min_aspect_ratio to optional, move rotation and pad before random resized crop

* fix

* Trigger CI

* Trigger CI

* Trigger CI

* fix default values

* Trigger CI

reminisce committed:
Address cr

xinyu-intel committed:
Fix axis Bug in MKLDNN Softmax (#11335)

* add softmax imporvement

* reuse CheckAxis code

* update comment

* add tests with negative axis

azai91 committed:
[MXNET-498] Test MKLDNN backward operators  (#11232)

* add act backwards test

* use only verifyfn template

* fix param name

* update number of inputs

* fix assertion for act backwards

* limit rand num range

* change to assert

* wait to read on correct vector

* add writeinplace test

* fix params

* add copy backwards test

* add missing fixture

* fix lint

* add sum backwards verify

* use correct num of inputs for sum backwards

* switch input / output

* wait for both outputs

* limit input/output

* limit input/outputs for relu/sum

* fix var source

* reorder backwards if view

* add another entry to reqs in ttest

* uncomment write in place sumbackwards

* refactor testunary and testbinary into testop

* remove special testbackwardsop and use testop

* fill reqs vector with num of outputs

* change req size to num outputs

* create mulitple output ndarrays

* wait for all outputs

* remove unused comments

* remove redundant VerifyCopyResult method

* remove redundant VerifySumResult

* remove unused var

* use only InitDefaultArray

* move MKLDNNSum near copy test

* use fallback compute for backwards sum

* fix verifydefmem test

* fix lint

* move MKLDNNSum test back to bottom

piiswrong committed:
Static alloc for hybridblock (#11320)

* Revert "Revert "Static alloc for hybridblock (#11313)" (#11318)"

This reverts commit 9b8eb567273d9521c639ea783e1117fe09cb3d64.

* fix

anirudh2290 committed:
Fix save load doc (#11345)

haojin2 committed:
[MXNET-404] elemwise_add/sub between rsp and rsp on GPU (#11179)

* Support for elemwise_add/sub between rsp and rsp on GPU

* add extra test coverage for inplace cases

wenyangchu committed:
Add seed_aug parameter for ImageRecordItr to fix random seed for default augmentation (#11247)

* add seed_aug parameter for ImageRecordItr to fix random seed for augmentation

* remove white space

* add test

* fix test

* improve according to review: using dmlc::optional and has_value()

* missing header

* change data type and way to get value

reminisce committed:
Fix lint

reminisce committed:
Fix storage type vector

reminisce committed:
Temporarily disable test_gluon_trainer.test_trainer_reset_kv
